DOC: added note to docstring of numpy.savez

GuyInALabCoat opened the pull request:
Note indicates that when used with dictionaries, savez will remove 
leading "/" in dictionary keys.
See #9265

GuyInALabCoat committed:
DOC: added note to docstring of numpy.savez

Note indicates that when used with dictionaries, savez will remove
leading "/" in dictionary keys.
See #9265

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -561,6 +561,11 @@ def savez(file, *args, **kwds):
+    When saving dictionaries, avoid using keys that begin with "/", as the
+    first slash will be removed by the python archive function. If they 
+    must be used, prepend another character ahead of the slash to ensure
+    that the "/" will not be removed.
```
I think we should be more specific, and mention that keys are stored as zip file names

eric-wieser replied:
So if variable names contain `..`, you're in for a bad time too. Surprisingly, it seems that the reserved windows file name (like `nul`) aren't actually an issue here.

GuyInALabCoat replied:
Thank you for your input. Do you think we should state:
```
When saving dictionaries, the dictionary keys are themselves are stored as zip file names 
in the zip directory. Therefore one should avoid using key strings that are not valid file names. 

Examples include strings that begin with "/", or contain ".." or "./", such as
'/test/tmp.txt', './test/tmp.txt', or test/../tmp.txt'

Should keys that include such patterns be required, they must be properly encapsulated and or
escaped by using either double quotes or backslashes where necessary. Note that any 
backslashes will be doubled in the returned keys. For example:

np.savez('tmp', **{'\./test/\.\./tmp.txt': np.array([1, 2])})

will return ['\\./test/\\.\\./tmp.txt']
```

GuyInALabCoat committed:
DOC: edited numpy.savez docstring note to be more specific

Note states that dictionary keys are stored as file names and may
undergo unintended changes when they are not recognized as valid
file names. In such cases the keys must be properly encapsulated
and escaped to avoid being changed.

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -561,6 +561,20 @@ def savez(file, *args, **kwds):
+    Should keys that include such patterns be required, they must be 
+    properly encapsulated and or escaped using either double quotes or
+    backslashes where necessary. Note that backslashes will be 
+    doubled in the returned keys.
```
This isn't good advice - this often won't be enough, especially on windows where backslash is a path separator. Realistically, we have two options here:

* Tell the user that they're not choosing string names, but (relative?) file names
* percent-encode the strings within `savez` and `loadz`, such that the user thinks they are working with raw strings. This might have compatibility ramifications

charris reviewed (changes requested):
Note the comment by @eric-wieser .

stefanv committed:
Tidy up savez keys-as-file-names language

stefanv committed:
Use filename (vs file name) consistently

mattip commented:
Thanks @GuyInALabCoat
